DBCON-257: Troubleshooting  DB Connector Guides in Public Documentation (#1404)
mulesoft/docs-connectors: db/1.9/modules/ROOT/pages/database-connector-troubleshooting.adoc

= Troubleshoot Database Connector

> To troubleshoot Anypoint Connector for Database (Database Connector), review the following sections for performing general troubleshooting, addressing specific common errors, and interpreting commonly thrown messages.

Reviewer: OH I SEE NOW...this is boilerplate. OK, sorry about the comment in that other edit, about making these specific. But I do think that this wording is better for the template:
```suggestion
To most efficiently troubleshoot Anypoint Connector for Database (Database Connector), become familiar with the recommendations for performing general troubleshooting, addressing specific common errors, and interpreting commonly thrown messages.
```

Database Connector follows the standards and guidelines detailed by the JDBC library (working with Java SE JDK 1.8).

== Enable Verbose Logging

> Sometimes you are not sure why the application is failing, and you require a better understanding of what is going on in the interaction with the database. You can get more detailed logs on the Mule application side by enabling verbose logging for Database Connector. +

Reviewer: copyedit:
```suggestion
Sometimes you are not sure why the application is failing, and you require a better understanding of what is going on in its interaction with the database. You can get more detailed logs on the Mule application side by temporarily enabling verbose logging for Database Connector. +
```
More concise:
```suggestion
To get a better understanding of why an application's interaction with the database is failing, temporarily enable verbose logging for Database Connector. +
```

> There are two options to enable verbose logging, one is to manually change the `log4j2.xml` file, or through Anypoint Runtime Manager if you are using CloudHub APIs. +

Reviewer: GLOBALLY avoid using the false sentence subject "There is" or "There are"):
```suggestion
Choose one of two options to enable verbose logging: manually change the `log4j2.xml` file or via Anypoint Runtime Manager if you are using CloudHub APIs. +
```
Do we need to specify exactly what one does in ARM to enable logging?

> Remember to always disable the enhanced verbosity after troubleshooting since it affects your Mule application performance.

Reviewer: copyedit (GLOBALLY reevaluate using "since": it is more precise to use either "after" or "because," depending on your intent:
```suggestion
Remember to always disable the enhanced verbosity after troubleshooting, because it affects your Mule application performance.
```
Alternative:
```suggestion
Whichever option you choose, always disable enhanced verbosity after troubleshooting, because the overhead it requires reduces application performance.
```

To enable verbose logging in the configuration file:

. Access Anypoint Studio and navigate to the *Package Explorer* view.
. Open your application's project name.
. Open the *src/main/resources* path folder.
. Open the `log4j2.xml` file inside the folder.

> . Add an `<AsyncLogger>` tag inside the `<Loggers>` tag, for example:

Reviewer: ```suggestion
. Add an `<AsyncLogger>` tag inside the `<Loggers>` tag:
```

+
[source,xml,linenums]
----
			<Loggers>
				...
				<AsyncLogger name="org.mule.extension.db" level="DEBUG"/>
				...
			</Loggers>
----
[start=6]

> . Save all your application changes.

Reviewer: ```suggestion
. Save your application changes.
```

. Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*.

To enable verbose logging in Anypoint Runtime Manager:

. Access Anypoint Runtime Manager.
. Navigate to the API configuration.

> . In the *Logging* sub-tab, add the `org.mule.extension.db` package at DEBUG level.

Reviewer: ```suggestion
. In the *Logging* subtab, add the `org.mule.extension.db` package at DEBUG level.
```

. Apply the changes.

== Troubleshoot Pooling Profiles Connections Making the Flow Hang
Using a pooling profile that provides fewer connections than the flow requires can make the entire flow hang.

To resolve this issue:

> One solution is to increase the size of the pool. However, sometimes that's not possible, so another solution is to use Anypoint Connector for Virtual Machine (VM Connector) to shorten the life of the connection to match the life of a VM message. This way, every call to the *Stored procedure* operation is executed, its result is consumed, and the connection associated with the operation is closed and returned to the pool.

Reviewer: ```suggestion
One solution is to increase the size of the pool. When that's not possible, use Anypoint Connector for Virtual Machine (VM Connector) to shorten the life of the connection to match the life of a VM message. This way, every call to the *Stored procedure* operation is executed, its result is consumed, and the connection associated with the operation is closed and returned to the pool.
```

	</flow>
----

> == Troubleshoot Snowflake JDBC Driver that Throws an Exception for a Stored Procedure

Reviewer: ```suggestion
== Troubleshoot Snowflake JDBC Driver That Throws an Exception for a Stored Procedure
```

Some drivers have their own specifications when calling stored procedures. Running the application when you set the callable statement with curly brackets (`{` and `}`), for example, `{call doubleMyInt(:myInt)}`, causes the Snowflake JDBC driver to throw the following exception:

`net.snowflake.client.jdbc.SnowflakeSQLException: Unsupported feature 'call'`

To resolve this issue:

For Snowflake JDBC driver, set the callable statement without the curly brackets `{` and `}`, for example:

`call doubleMyInt(:myInt)`

== Understand Common Throws

Here is a list of common throws messages and how to interpret them.

* DB:BAD_SQL_SYNTAX

>  The provided SQL query has a bad syntax.

Reviewer: ```suggestion
 The provided SQL query contains invalid syntax.
```

* DB:CANNOT_LOAD_DRIVER

>  Database Connector can not load the JDBC driver.

Reviewer: ```suggestion
 Database Connector cannot load the JDBC driver.
```

* DB:CANNOT_REACH

>  Database Connector can not establish a connection with the RDBMS.

Reviewer: ```suggestion
 Database Connector cannot establish a connection with the RDBMS.
```

* DB:CONNECTIVITY

>  A connectivity issue with the database that occurs on badly configured connections, no database response while waiting results or fetching rows, lost of database connection, and similars.

Reviewer: Please review and reword; this is not grammatical.
Here is my best guess:
```suggestion
 Any of several possible connectivity issues, including poorly configured connections, lack of database response to queries, loss of database connection, and so on.
```

>  Any of several possible connectivity issues, including poorly configured connections, lack of database response waiting results or fetching rows, loss of database connection, and so on.

Reviewer 2: ```suggestion
 Any of several possible connectivity issues, including poorly configured connections, lack of a database response while awaiting results or fetching rows, loss of database connection, and so on.
```

Reviewer 2: ? I'm not sure what this is meant to say: ...lack of database response waiting results or fetching rows,...

* DB:INVALID_DATABASE

 The connection to the RDBMS is established but the database doesn't exist.

* DB:INVALID_CREDENTIALS

 The database rejects the provided credentials.

* DB:QUERY_EXECUTION

>  There is an error executing the query.

Reviewer: This just repeats the error message; is there any value to add here, maybe some possible causes?

* DB:RETRY_EXHAUSTED

>  All reconnection attempts failed.

Reviewer: Is this a maximum specified number? If so, should we reword to include that?
